src/main/java/org/broadinstitute/hellbender/engine/MultiVariantDataSource.java

```diff
@@ -0,0 +1,408 @@
+            mergedHeader = headers.size() > 1 ?
+                    new VCFHeader(VCFUtils.smartMergeHeaders(headers, true)) :
+                    headers.get(0);
+        }
+        return mergedHeader;
+    }
+
+    /**
+     * Permanently close this data source, invalidating any open iteration over it, and making it invalid for future
+     * iterations and queries.
+     */
+    @Override
+    public void close() {
+        validateFeatureSources();
+        featureDataSources.forEach(dataSource -> dataSource.close());
+    }
+
+    /**
+     * Cross-validate the sequence records from the sequence dictionaries from all data sources, since its possible
+     * for any given pair to have conflicts that are unique to that pair. GATKTool only validates individual feature
+     * dictionaries against the reference dictionary, so we need to cross-validate them against each other here
+     * before we merge them together.
+     */
+    private void validateAllSequenceDictionaries() {
+        Map<String, FeatureDataSource<VariantContext>> contigMap = new HashMap<>();
+        featureDataSources.forEach(
+            ds -> {
+                ds.getSequenceDictionary().getSequences().forEach(
+                    sourceSequence -> {
+                        String sourceName = sourceSequence.getSequenceName();
+                        FeatureDataSource<VariantContext> targetDataSource = contigMap.getOrDefault(sourceName, null);
+                        if (targetDataSource != null) {
+                            SAMSequenceDictionary targetDictionary = targetDataSource.getSequenceDictionary();
+                            SAMSequenceRecord targetSequence = targetDictionary.getSequence(sourceName);
+                            validateSequenceDictionaryRecords(
+                                    ds.getName(),
+                                    ds.getSequenceDictionary(),
+                                    sourceSequence,
+                                    targetDataSource.getName(),
+                                    targetDictionary,
+                                    targetSequence);
+                        } else {
+                            contigMap.put(sourceName, ds);
+                        }
+                    }
+                );
+            }
+        );
+    }
+
+    // Cross validate the length and md5 for a pair of sequence records and throw if they're not the same.
+    private void validateSequenceDictionaryRecords(
+            final String sourceDataSourceName,
+            final SAMSequenceDictionary sourceDictionary,
+            final SAMSequenceRecord sourceSequence,
+            final String targetDataSourceName,
+            final SAMSequenceDictionary targetDictionary,
+            final SAMSequenceRecord targetSequence)
+    {
+        if (targetSequence.getSequenceLength() != sourceSequence.getSequenceLength()) {
```

Comment:
I think you want to call into the existing `SequenceDictionaryUtils.sequenceRecordsAreEquivalent()` method to do the name and length checks (since it handles some edge cases), and then do the MD5 check on top of it (although checking the MD5s is stricter than we usually are).